Async data loading

rollaball opened the pull request:
* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Data can be fetched asynchronously data when a property `async` is set to `true`


* **What is the current behavior?** (You can also link to an open issue here)


* **What is the new behavior (if this is a feature change)?**



* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)



* **Other information**:
Issue: #317

rollaball commented:
@mathisonian 
is this good?
added 2 properties to data `async` and `initialValue`.
eg: `[data name:'myAsyncData' source:'url' async:true initialValue:'[]']`

mathisonian commented:
Sweet that API looks good @rollaball 👍 

I haven't tested the changes locally yet, but the code updates so far look good. Is there any way to do it without adding the `axios` dependency?

rollaball commented:
@mathisonian , I am not sure how we can do this without package such as axios. What do you recommend?

mathisonian commented:
Fetch seems like a good solution 👍. This won't work in IE 11, so we should add a note to the docs about that. Users can include their own fetch polyfill if they need async data + IE11 support.

mathisonian commented:
I'll do a full review of this today.

mathisonian commented on packages/idyll-document/src/utils/index.js:
```diff
@@ -238,12 +239,48 @@ export const filterIdyllProps = (props, filterInjected) => {
+            if (sourceValue.endsWith('.csv')) {
+              return res
+                .text()
+                .then(resString => parse(resString))
```
We should use the same options as the parsing that happens in the CLI so there aren't unexpected differences. There we call `parse(inputString, { cast: true, columns: true })`

rollaball replied:
updated

mathisonian commented on packages/idyll-document/src/utils/index.js:
```diff
@@ -238,12 +239,48 @@ export const filterIdyllProps = (props, filterInjected) => {
+
+      let dataPromise = new Promise(res => res(initialValue));
+
+      if (typeof fetch !== 'undefined') {
```
Can we print a warning if this check fails?

rollaball replied:
updated

rollaball commented:
@mathisonian , I have updated the code and added tests. The tests seem to be failing for some reason as it is unable to fetch the repo in sometimes so could you re-run it once.

mathisonian commented:
Awesome work @rollaball ! 🎉
